text/0026-block-version-based-protocol-evolution.md

```diff
@@ -0,0 +1,228 @@
+- Feature Name: Block-Version-based Protocol Evolution
+- Start Date: 2022-02-05
+- MCIP PR: [mobilecoinfoundation/mcips#0026](https://github.com/mobilecoinfoundation/mcips/pull/0026)
+- MobileCoin Epic: None
+
+# Summary
+[summary]: #summary
+
+Since inception, MobileCoin has not yet made any changes to the ledger format.
+
+This document proposes a general approach to rolling out changes to the ledger format
+efficiently while carefully maintaining backwards compatibility.
+
+# Motivation
+[motivation]: #motivation
+
+Conventional web applications today are usually developed using data interchange
+formats that support schema evolution with forward and backwards compatibility.
+A good example of this is protobuf -- protobuf allows that new fields can be
+added to schema, and new programs will be able to deserialize old payloads, and
+old programs will be able to deserialize the new payloads (ignoring the new fields).
+
+As long as changes are strictly additive then they can be rolled out without
+breaking old clients. Eventually the old clients are sunsetted and only the new
+clients are supported.
+
+In blockchains, this is made more complicated for two reasons:
+* Older data payloads exist forever in the blockchain and clients have to be able to handle them forever
+* Frequently, clients and nodes need to agree about the hash of of a piece of data, so that
+  clients that sync the blockchain compute the correct block id, and clients that submit merkle proofs
+  of membership compute correct proofs.
+
+Because of the need for everyone to compute hashes correctly, it may not be enough for
+an old client to successfully deserialize a new payload by ignoring the fields it doesn't
+understand -- if those fields are ignored, then it will compute a different hash for
+this object from the new clients.
+
+To make this concrete, this means that if we want to roll out changes that add new
+features to `TxOut`'s, the new clients and old clients probably cannot coexist the way
+they might in a more conventional application.
+
+If a new client sends a `TxOut` to a user using an old client, which contains a `TxOut`
+feature that they don't recognize, then:
+1. They may fail to compute the correct block id
+2. If they submit a transaction spending the TxOut, the merkle proof of membership
+   may be wrong if the hash doesn't incorporate the new field.
```

Comment:
This is a good criticism of the dual-rule blocks approach, that I sadly did not foresee.

Reply:
@garbageslam What if membership proofs just hashed the core TxOut components? I.e. the onetime address and amount commitment. Then you can change memo fields arbitrarily without old clients needing to know about it when constructing a membership proof for the old enclave in a 'dual-rule' block. At the very least, you'd only need to do single-rule transitions on the rare occasion that core TxOut components change (if ever).

Reply:
> This is a good criticism of the dual-rule blocks approach, that I sadly did not foresee.

Yeah I didn't really foresee this either. I saw pieces of it but not the whole thing. I think James saw the problem a while ago and it wasn't until now that we had a clear plan how to fix it.

Reply:
> At the very least, you'd only need to do single-rule transitions on the rare occasion that core TxOut components change (if ever).

I think the simplest way to make this work without this kind of "dual rule" stuff would have been if:

* Instead of mc-crypto-digestible, we hash raw protobuf payloads, and clients leave all the fields in there for this hash even if they don't recognize some of them.
* Because protobuf serializations are not canonical, we would have to either make a "canonical protobuf" serializer, or, a routine that takes protobufs and makes them canonical.

It's a little late in the game for that. Making a protobuf serializer is really complicated, and we didn't really have the resources to try to do that, esp. that would support multiple languages, when we had only a few engineers. The `mc-crypto-digestible` thing was a lot easier to put together. We could also have tried to make a routine that takes serialized protobuf payloads and reorders the fields to make them canonical, but that also would be pretty complicated.

Reply:
> I think it still may break clients that are syncing the whole ledger, because if they are on an old block version, and TxOut's appear that have new fields, they may compute the wrong block id and then reject the block. So that would instantly force those older clients to update or be broken. So I don't think that's a complete solution unfortunately, otherwise yeah it would have been my preference to do it that way

Yeah anyone syncing the ledger needs a full and updated codebase. I think it would be wasted effort to try to make universal backward compatibility across forks.

The goals of dual-rule blocks are: 1) prevent txs from failing if they are submitted right before a transition point, 2) let 'simple' codebases like wallets only worry about one enclave at a time with loose upgrade requirements around forks (i.e. wallets that aren't syncing the entire chain).

Reply:
> Yeah anyone syncing the ledger needs a full and updated codebase. I think it would be wasted effort to try to make universal backward compatibility of old clients across forks.

I think the model we want to have is like,
* Users are expected to update every X days. If you have a phone app, it will prompt you to approve updates periodically. If you are running a point-of-sale terminal, you may have to do scheduled maintenance. This is all normal.
* Users expect that as long as they do the upgrades on schedule, the system always works.

Suppose that e.g. a grocery store decided to do a trial run of MOB point-of-sale terminals, which are syncing the ledger probably, using full-service or mobilecoind let's say.

If the terminal is old and the customer has new versions of our software, and the terminal breaks (because it can't spend a TxOut with a new field), that's not a good experience for the merchant. Perhaps they can fix it by doing a update but it's not the experience they have with VISA for instance. However, if they update aggressively (and before their customers), then their customers are the ones having the unspendable TxOut's, if they have to process a refund or something. So it's a bit of a catch 22, someone is going to have a broken transaction and have to suddenly do an update faster than expected.

I think we'd really like to be able to ship clients that are guaranteed to work for their life-time and not break them prematurely, no matter if they are desktop wallets or mobile wallets. I don't personally think this is wasted effort, as long as we figure out a system that works and follow it, then everyone will have a good user experience.

Reply:
> I think we'd really like to be able to ship clients that are guaranteed to work for their life-time and not break them prematurely, no matter if they are desktop wallets or mobile wallets.

This can only be achieved for scanning the ledger for owned outputs (bare-minimum balance-checking where the only things you look at are onetime address, amount commitments, encoded amounts, output pubkeys, and key images), assuming you don't transition to new crypto for those pieces (e.g. Seraphis). Spending to a new enclave always requires an updated client, since you need to sign all tx data which means knowing all the current formats/etc. Reading miscellaneous new memos also always requires an updated client (obviously).

Reply:
Yes but if we have a policy like, we distribute updates every X days that trust both current MRENCLAVE and next MRENCLAVE, and only after `X` days do we deploy the new enclaves, we can also navigate that without ever breaking clients that are less than `X` days old.